lib/broadway/options.ex

```diff
@@ -218,10 +218,28 @@ defmodule Broadway.Options do
               """
             ],
             batch_size: [
-              type: :pos_integer,
+              type: {:custom, __MODULE__, :validate_batch_size, []},
               default: 100,
               doc: """
-              The size of the generated batches.
+              The size of the generated batches. Default value is `100`.
+              It can also be tuple of {:fun/2, init_size}.  The function must have an arity of 2
+              with `Broadway.Message` and last calculated `acc` returned from it.  `init_size` is provided
+              as the second parameter on first call.
+              The function must return either {:emit, acc} to indicate a full batch,
+              or {:cont, acc} to continue.
+              Default batch size is actually converted to this:
+              `{fn _message, remained ->
+                 remained = remained - 1
+                 if(remained == 0, do: {:emit, batch_size}, else: {:cont, remained})
+               end, batch_size}`
+              """
```

Comment:
```suggestion
              The size of the generated batches. Default value is `100`.

              It is typically an integer but it can also be tuple of `{fun, init_acc}`
              where `fun` receives two arguments: a `Broadway.Message` and
              an `acc`.  The function must return either `{:emit, acc}` to indicate
              all batched messages must be emitted or `{:cont, acc}` to continue
              batching. `init_acc` is the initial accumulator used on the first call.

              You can consider that setting the accumulator to an integer is the
              equivalent to custom batching function of:

                  {fn _message, remained ->
                     remained = remained - 1
                     if(remained == 0, do: {:emit, batch_size}, else: {:cont, remained})
                   end, batch_size}

              """
```